pkg/storegateway/indexheader/binary_reader.go

```diff
@@ -947,30 +950,67 @@ func (b realByteSlice) Sub(start, end int) index.ByteSlice {
 // found entry. If f returns an error it stops decoding and returns the received error.
 // TODO: use the upstreamed version from Prometheus tsdb/index package when available.
 // TODO: see https://github.com/prometheus/prometheus/pull/11535
-func readPostingsOffsetTable(bs index.ByteSlice, off uint64, f func(labels.Label, uint64, int) error) error {
+func readOffsetTable[T any](bs index.ByteSlice, off uint64, read func(d *encoding.Decbuf) (T, error), f func(T, uint64, int) error) error {
 	d := encoding.NewDecbufAt(bs, int(off), castagnoliTable)
 	startLen := d.Len()
 	cnt := d.Be32()
 
 	for d.Err() == nil && d.Len() > 0 && cnt > 0 {
 		offsetPos := startLen - d.Len()
 
-		if keyCount := d.Uvarint(); keyCount != 2 {
-			return errors.Errorf("unexpected key length for posting table %d", keyCount)
+		item, err := read(&d)
+		if err != nil {
+			return err
 		}
-
-		var lbl labels.Label
-		lbl.Name = d.UvarintStr()
-		lbl.Value = d.UvarintStr()
-
 		o := d.Uvarint64()
 		if d.Err() != nil {
 			break
 		}
-		if err := f(lbl, o, offsetPos); err != nil {
+		if err := f(item, o, offsetPos); err != nil {
 			return err
 		}
 		cnt--
 	}
 	return d.Err()
 }
+
+func postingsOffsetTableReader(d *encoding.Decbuf) (labels.Label, error) {
+	if keyCount := d.Uvarint(); keyCount != 2 {
+		return labels.Label{}, errors.Errorf("unexpected key length for posting table %d", keyCount)
+	}
+	var lbl labels.Label
+	lbl.Name = d.UvarintStr()
+	lbl.Value = d.UvarintStr()
+	return lbl, nil
+}
+
+// yoloPostingsOffsetTableReader reads the postings table returning a yoloed label, which can be used for comparison
+// without allocating a new string (since in most of the cases, we just want to see if the label name has changed).
+func yoloPostingsOffsetTableReader(d *encoding.Decbuf) (yoloLabel, error) {
```

Comment:
Since this is all new code, can we use a different prefix than "yolo"? Maybe something like "mapped", "unsafe", "shared", etc.?

Reply:
What if we also rename `yoloLabel` to `unsafeLabel` and its `name()` and `value()` functions to `safeName()` and `safeValue()` to clarify they make a copy and the returned string is safe?

Reply:
Done in https://github.com/grafana/mimir/pull/3397/commits/0a6f7875d7dc31cc1b4775f18ddef94ee2dc9b5a